projects/packages/backup-helper-script-manager/src/class-helper-script-manager.php

```diff
@@ -0,0 +1,584 @@
+<?php
+/**
+ * The Jetpack Backup Helper Script Manager class.
+ *
+ * @package automattic/jetpack-backup
+ */
+
+namespace Automattic\Jetpack\Backup;
+
+use Exception;
+
+/**
+ * Helper_Script_Manager manages installation, deletion and cleanup of Helper Scripts
+ * to assist with backing up Jetpack Sites.
+ *
+ * Does *not* use WP_Filesystem, because the helper script is something that we'll call over HTTP, so we don't want it
+ * to end up on an FTP server somewhere. Also, PHP provides us with better error reporting than WP_Filesystem.
```

Comment:
My understanding is that `WP_Filesystem` is a mechanism that allows one to serve various resources (e.g. media files) from an FTP/SFTP server (given the fact that subclasses `WP_Filesystem_FTPext`, `WP_Filesystem_ftpsocket`, `WP_Filesystem_SSH2` exist, in addition to `WP_Filesystem_Direct` which is just a thin layer over the likes of `file_put_contents()`).

Our helper script is not a resource, but rather an executable PHP file which we intend to be able to access over HTTP after uploading it, so we *do not* want it to accidentally end up on an FTP server somewhere where it won't be accessible to us.

Something like a WordPress plugin editor doesn't use `WP_Filesystem` either, I think it's for the same reason:

https://github.com/WordPress/WordPress/blob/a2bcc36278ea932406a3e0044a6d9f96c8493c10/wp-admin/plugin-editor.php#L167-L171

Reply:
That's not a good example, as it's using `file_get_contents` (a read operation, if source files couldn't be read then even `require` would break) rather than a write operation like `file_put_contents`.

Reply:
https://developer.wordpress.org/apis/filesystem/#purpose may be a useful explanation of what `WP_Filesystem` is for.

Reply:
> That's not a good example, as it's using file_get_contents (a read operation, if source files couldn't be read then even require would break) rather than a write operation like file_put_contents.

I think one is supposed to fetch contents with `WP_Filesystem_Base::get_contents()`.

> https://developer.wordpress.org/apis/filesystem/#purpose may be a useful explanation of what WP_Filesystem is for.

Thanks, TIL, but it does not apply to us in this case, because if there are permissions issues between the webserver's user and the FTP/SSH user, then we'll just install the helper script and do a backup/restore using FTP/SSH credentials (that we collect ourselves), without using `WP_Filesystem` in any way.

Also, if we can't write that helper script somewhere (due to writes being inaccessible to the webserver's user, or for other reasons), we want to know about it (in the form of an error response), instead of having that helper script silently uploaded via FTP/SFTP, so that we could fallback to a backup/restore using credentials.

Reply:
> if there are permissions issues between the webserver's user and the FTP/SSH user, then we'll just install the helper script and do a backup/restore using FTP/SSH credentials (that we collect ourselves), without using WP_Filesystem in any way.
>
> Also, if we can't write that helper script somewhere (due to writes being inaccessible to the webserver's user, or for other reasons), we want to know about it (in the form of an error response), instead of having that helper script silently uploaded via FTP/SFTP, so that we could fallback to a backup/restore using credentials.

Thanks, that would be a good explanation to put in the comment instead. 🙂

Reply:
Thanks, updated comment in a77d279.